In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [50]:
df=pd.read_csv(r"C:\Users\-----\Downloads\Nassau_Candy_Distributor.csv")
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Country/Region,City,State/Province,Postal Code,Division,Region,Product ID,Product Name,Sales,Units,Gross Profit,Cost
0,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.50,2,4.22,2.28
1,2,US-2021-112326-CHO-TRI-54000,04-01-2024,01-07-2026,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-TRI-54000,Wonka Bar - Triple Dazzle Caramel,7.50,2,4.90,2.60
2,3,US-2021-112326-CHO-NUT-13000,04-01-2024,01-07-2026,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-NUT-13000,Wonka Bar - Nutty Crunch Surprise,10.47,3,7.47,3.00
3,4,US-2021-112326-CHO-SCR-58000,04-01-2024,01-07-2026,Standard Class,112326,United States,Naperville,Illinois,60540,Chocolate,Interior,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.80,3,7.50,3.30
4,5,US-2021-141817-CHO-TRI-54000,05-01-2024,05-07-2026,Standard Class,141817,United States,Philadelphia,Pennsylvania,19143,Chocolate,Atlantic,CHO-TRI-54000,Wonka Bar - Triple Dazzle Caramel,11.25,3,7.35,3.90


In [15]:
df1=pd.read_csv(r"C:\Users\-----\Downloads\Products and Factories Correlation.csv")
df1.head()

,Division,Product Name,Factory
0,Chocolate,Wonka Bar - Nutty Crunch Surprise,Lot's O' Nuts
1,Chocolate,Wonka Bar - Fudge Mallows,Lot's O' Nuts
2,Chocolate,Wonka Bar -Scrumdiddlyumptious,Lot's O' Nuts
3,Chocolate,Wonka Bar - Milk Chocolate,Wicked Choccy's
4,Chocolate,Wonka Bar - Triple Dazzle Caramel,Wicked Choccy's


Data Cleaning & Validation
• Validate cost and sales values
• Remove zero-sales or invalid profit records
• Handle missing unit values
• Standardize product and division labels

In [18]:
df.isnull().sum()

Row ID            0
Order ID          0
Order Date        0
Ship Date         0
Ship Mode         0
Customer ID       0
Country/Region    0
City              0
State/Province    0
Postal Code       0
Division          0
Region            0
Product ID        0
Product Name      0
Sales             0
Units             0
Gross Profit      0
Cost              0
Gross Margin      0
dtype: int64

In [60]:
df.shape

(10194, 18)

In [9]:
df.describe()

,Row ID,Customer ID,Sales,Units,Gross Profit,Cost
count,10194.000000,10194.000000,10194.000000,10194.000000,10194.000000,10194.000000
mean,5097.500000,134468.961154,13.908537,3.791838,9.166451,4.742087
std,2942.898656,20231.483007,11.341020,2.228317,6.643740,5.061647
min,1.000000,100006.000000,1.250000,1.000000,0.250000,0.600000
25%,2549.250000,117212.000000,7.200000,2.000000,4.900000,2.400000
50%,5097.500000,133550.000000,10.800000,3.000000,7.470000,3.600000
75%,7645.750000,152051.000000,18.000000,5.000000,12.250000,5.700000
max,10194.000000,192314.000000,260.000000,14.000000,130.000000,130.000000


In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10194 entries, 0 to 10193
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Row ID          10194 non-null  int64  
 1   Order ID        10194 non-null  object 
 2   Order Date      10194 non-null  object 
 3   Ship Date       10194 non-null  object 
 4   Ship Mode       10194 non-null  object 
 5   Customer ID     10194 non-null  int64  
 6   Country/Region  10194 non-null  object 
 7   City            10194 non-null  object 
 8   State/Province  10194 non-null  object 
 9   Postal Code     10194 non-null  object 
 10  Division        10194 non-null  object 
 11  Region          10194 non-null  object 
 12  Product ID      10194 non-null  object 
 13  Product Name    10194 non-null  object 
 14  Sales           10194 non-null  float64
 15  Units           10194 non-null  int64  
 16  Gross Profit    10194 non-null  float64
 17  Cost            10194 non-null 

In [71]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Country/Region', 'City', 'State/Province',
       'Postal Code', 'Division', 'Region', 'Product ID', 'Product Name',
       'Sales', 'Units', 'Gross Profit', 'Cost', 'Profit_%', 'Profit Status',
       'Gross_Margin'],
      dtype='object')

In [75]:
df[df['Units']==0]

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Country/Region,City,State/Province,Postal Code,Division,Region,Product ID,Product Name,Sales,Units,Gross Profit,Cost,Gross_Margin,Profit_%


In [74]:
df[['Sales','Cost']].isnull().sum()

Sales    0
Cost     0
dtype: int64

In [92]:
df['Product Name'].unique()

array(['Wonka Bar - Milk Chocolate', 'Wonka Bar - Triple Dazzle Caramel',
       'Wonka Bar - Nutty Crunch Surprise',
       'Wonka Bar - Scrumdiddlyumptious', 'Wonka Bar - Fudge Mallows',
       'Wonka Gum', 'Kazookles', 'Lickable Wallpaper',
       'Fizzy Lifting Drinks', 'Laffy Taffy', 'Sweet Arts', 'Nerds',
       'Hair Toffee', 'Everlasting Gobstopper', 'Fun Dip'], dtype=object)

In [77]:
df['Division'].unique()

array(['Chocolate', 'Other', 'Sugar'], dtype=object)

In [ ]:
## Standardize product and division labels

In [101]:
## Standardize product and division labels
df['Product Name']=df['Product Name'].replace({'SweeTARTS':'Sweet arts'})
df['Product Name']=df['Product Name'].str.strip().str.title()
df['Product Name']=df['Product Name'].replace({'Wonka Bar -Scrumdiddlyumptious':'Wonka Bar - Scrumdiddlyumptious'})
df['City']=df['City'].str.strip().str.title()
df['State/Province']=df['State/Province'].str.strip().str.title()

## Standardized the values in ‘Product Name’, ‘Division’, ‘City’, and ‘State/Province’ to ensure consistency and accuracy across the dataset.

In [102]:
df['State/Province'].unique()

array(['Texas', 'Illinois', 'Pennsylvania', 'Kentucky', 'Georgia',
       'California', 'Virginia', 'Delaware', 'South Carolina', 'Ohio',
       'Louisiana', 'Oregon', 'Arizona', 'Arkansas', 'Michigan',
       'Tennessee', 'Florida', 'Ontario', 'Indiana', 'Nevada',
       'South Dakota', 'New York', 'Wisconsin', 'Washington',
       'New Jersey', 'Missouri', 'North Carolina', 'Colorado', 'Alberta',
       'Utah', 'Minnesota', 'Mississippi', 'Iowa', 'New Mexico',
       'Massachusetts', 'Alabama', 'Idaho', 'Montana', 'Maryland',
       'Connecticut', 'New Hampshire', 'British Columbia', 'Quebec',
       'Nova Scotia', 'Oklahoma', 'Nebraska', 'Maine', 'Kansas',
       'Rhode Island', 'Newfoundland And Labrador', 'New Brunswick',
       'Prince Edward Island', 'District Of Columbia', 'Vermont',
       'Manitoba', 'Saskatchewan', 'Wyoming', 'North Dakota',
       'West Virginia'], dtype=object)

In [115]:
df['Order ID'].value_counts()
df['Order ID'].value_counts()[df['Order ID'].value_counts()==1].shape

(7176,)

## Problem Statement
Currently, the organization lacks visibility into:

#1. Which product lines deliver the highest gross margin
#2. Whether high-sales products are actually profitable
#3. How profitability varies across product divisions
#4. Which products represent margin risk

In [6]:
#1 Which product lines deliver the highest gross margin
df=df.groupby('Product Name')[['Units','Cost','Sales','Gross Profit']].sum()
Total_Units=df['Units'].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
df['Total Units']=((df['Units']/Total_Units)*100).round(2)
df.sort_values(by='Gross Margin', ascending=False).head(1)

,Units,Cost,Sales,Gross Profit,Gross Margin,Total Units
Product Name,,,,,,
Everlasting Gobstopper,13,26.0,130.0,104.0,80.0,0.03


### Observations: Which product lines deliver the highest gross margin
1. Everlasting Gobstopper (Product Line) has the highest gross margin of 80%, but it has sold only 0.03% units of the overall inventory.
2. As per the highest selling units, Wonka Bar - Nutty Crunch Surprise is the product that delivers a high gross margin. This product contributes 17.48% of the total units sold with a 71.35% gross margin.

In [ ]:
#2. Whether high-sales products are actually profitable
## Total of all rows
total_units=df['Units'].sum()
total_cost=df['Cost'].sum()
total_sales=df['Sales'].sum()
total_gross_profit=df['Gross Profit'].sum()

## Top 5 calculations
Overall_Gross_Margin=((df['Gross Profit'].sum()/df['Sales'].sum())*100).round(2)
top_5=df.groupby('Product Name')[['Units','Sales','Gross Profit','Gross Margin']].sum().sort_values(by='Sales', ascending=False).head(5)
df['Total Units %']=(top_5['Units']/total_units)*100
df['Total Sales %']=(top_5['Sales']/total_sales)*100
df['Gross Profit %']=(top_5['Gross Profit']/total_gross_profit)*100
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
top5_Gross_Margin=((top_5['Gross Profit'].sum()/top_5['Sales'].sum())*100).round(5)

## Top 5 Highest Selling SKUS Calculations
top5_units=top_5['Units'].sum()
top5_sales=top_5['Sales'].sum().round(2)
top5_profit=top_5['Gross Profit'].sum().round(2)
Top_5_Units_Contribution=((top5_units/total_units)*100).round(2)
Top_5_Sales_Contribution=((top5_sales/total_sales)*100).round(2)
Top_5_Profit=((top5_profit/total_gross_profit)*100).round(2)

## Print Statement
print("Top 5 Products Units Sold : ",top5_units)
print("Top 5 Products Units Contribution : ",Top_5_Units_Contribution,"%")
print("Top 5 Products Sales : ",top5_sales)
print("Top 5 Products Sales Contribution : ",Top_5_Sales_Contribution,"%")
print("Top 5 Products Gross Profit : ",top5_profit)
print("Top 5 Products Gross Profit Contribution : ",Top_5_Profit,"%")
print("Gross Margin Average of Top 5 Selling Products : " ,top5_Gross_Margin,"%")

#2 Observations: Whether high-sales products are actually profitable
1. Yes, based on the observations, the top 5 highest-selling SKUs contribute 95.06% of the total gross profit and 92.88% of the overall sales.
2. The average gross margin for the top 5 highest-selling SKUs is 67.45%, which is above the 60% margin bracket. This indicates that 96.43% of 
the units and 92.88% of the sales are generated with a 67.45% gross margin, which is a strong indicator of good profitability for the company.

In [ ]:
#3. How profitability varies across product divisions

In [15]:
def profit_status (x):
    if x > 60:
        return 'High Profit'
    elif x > 40:
        return 'Mid Profit'
    elif x > 20:
        return 'Low Profit'
    else:
        return 'Very low profit'

In [16]:
profitability=df.groupby('Division')[['Sales','Units','Cost','Gross Profit']].sum()
profitability['Profit_%']=((profitability['Gross Profit']/profitability['Sales'])*100).round(2)
profitability['Status']=profitability['Profit_%'].apply(profit_status)
profitability=profitability.sort_values(by='Units', ascending=False)
profitability

,Sales,Units,Cost,Gross Profit,Profit_%,Status
Division,,,,,,
Chocolate,131692.90,37275,42868.28,88824.62,67.45,High Profit
Other,9663.25,1242,5329.80,4333.45,44.84,Mid Profit
Sugar,427.48,137,142.75,284.73,66.61,High Profit


### Observations: How profitability varies across product divisions
1. The Chocolate category contributes 95.06% of the total gross profit, which clearly indicates that the Chocolate 
division dominates the company's overall profitability.
2. The Sugar and Other categories contribute 0.30% and 4.64% of the total gross profit, respectively, indicating that 
their impact on overall profitability is relatively small compared to the Chocolate division.

In [20]:
#4. Which products represent margin risk
df=df.groupby('Product Name')[['Units','Cost','Sales','Gross Profit']].sum()
Total_Units=df['Units'].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
df['Total Units']=((df['Units']/Total_Units)*100).round(2)
df.sort_values(by='Gross Margin', ascending=True).head(5)

,Units,Cost,Sales,Gross Profit,Gross Margin,Total Units
Product Name,,,,,,
Kazookles,371,1113.0,1205.75,92.75,7.69,0.96
Fun Dip,8,7.2,12.00,4.80,40.00,0.02
Nerds,10,8.0,15.00,7.00,46.67,0.03
Sweet arts,41,32.8,61.50,28.70,46.67,0.11
Lickable Wallpaper,393,3930.0,7860.00,3930.00,50.00,1.02


## Observations
1. Kazookles is the only product with margin risk with margin of 7.69% with sold over 370 units.
2. Fun Dip, SweeTARTS & Nerds are 3 products with profit margin below 50% except Kazookles 

Profitability Metric Calculation
For each product:
## Gross Margin (%)
## Profit per unit
## Total profit contribution

In [23]:
df['Per Unit Sale Price']=df['Sales']/df['Units']
df['Per Unit Cost']=df['Cost']/df['Units']
df['Profit Per Unit']=df['Per Unit Sale Price']-df['Per Unit Cost']
df['Gross Margin % Per Units']=(((df['Per Unit Sale Price']-df['Per Unit Cost'])/df['Per Unit Sale Price'])*100).round(2)
df

,Units,Cost,Sales,Gross Profit,Gross Margin,Total Units,Per Unit Sale Price,Per Unit Cost,Profit Per Unit,Gross Margin % Per Units
Product Name,,,,,,,,,,
Everlasting Gobstopper,13,26.00,130.00,104.00,80.00,0.03,10.00,2.00,8.00,80.00
Fizzy Lifting Drinks,21,31.50,78.75,47.25,60.00,0.05,3.75,1.50,2.25,60.00
Fun Dip,8,7.20,12.00,4.80,40.00,0.02,1.50,0.90,0.60,40.00
Hair Toffee,17,17.00,76.50,59.50,77.78,0.04,4.50,1.00,3.50,77.78
Kazookles,371,1113.00,1205.75,92.75,7.69,0.96,3.25,3.00,0.25,7.69
Laffy Taffy,27,20.25,53.73,33.48,62.31,0.07,1.99,0.75,1.24,62.31
Lickable Wallpaper,393,3930.00,7860.00,3930.00,50.00,1.02,20.00,10.00,10.00,50.00
Nerds,10,8.00,15.00,7.00,46.67,0.03,1.50,0.80,0.70,46.67
Sweet arts,41,32.80,61.50,28.70,46.67,0.11,1.50,0.80,0.70,46.67


In [ ]:
## Total profit contribution

In [29]:
#1. Gross Margin (%)
#1. Profit Per Unit
#3. Total Profit Contribution

df=df.groupby('Product Name')[['Units','Cost','Sales','Gross Profit']].sum()
Total_Units=df['Units'].sum()
Total_Profit=df['Gross Profit'].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
df['Units Contribution %']=((df['Units']/Total_Units)*100).round(2)
df['Profit Contribution %']=((df['Gross Profit']/Total_Profit)*100).round(2)
df['Profit Per Unit']=df['Gross Profit']/df['Units']
df.sort_values(by='Gross Margin', ascending=False)
top_profit=df.sort_values(by='Profit Contribution %',ascending=False)
top_profit

,Units,Cost,Sales,Gross Profit,Gross Margin,Units Contribution %,Profit Contribution %,Profit Per Unit
Product Name,,,,,,,,
Wonka Bar -Scrumdiddlyumptious,7743,8517.30,27874.80,19357.50,69.44,20.03,20.72,2.50
Wonka Bar - Triple Dazzle Caramel,7596,9874.80,28485.00,18610.20,65.33,19.65,19.92,2.45
Wonka Bar - Milk Chocolate,8267,9424.38,26867.75,17443.37,64.92,21.39,18.67,2.11
Wonka Bar - Nutty Crunch Surprise,6755,6755.00,23574.95,16819.95,71.35,17.48,18.00,2.49
Wonka Bar - Fudge Mallows,6914,8296.80,24890.40,16593.60,66.67,17.89,17.76,2.40
Lickable Wallpaper,393,3930.00,7860.00,3930.00,50.00,1.02,4.21,10.00
Wonka Gum,478,286.80,597.50,310.70,52.00,1.24,0.33,0.65
Everlasting Gobstopper,13,26.00,130.00,104.00,80.00,0.03,0.11,8.00
Kazookles,371,1113.00,1205.75,92.75,7.69,0.96,0.10,0.25


In [ ]:
## Product-Level Profitability Analysis
• Rank products by:
○ Gross profit
○ Gross margin

• Identify:
○ High-profit / high-margin products
○ High-sales / low-margin products
○ Low-sales / low-profit products

In [36]:
## Rank products by:
# Gross profit
ProductWise_Profit = df.groupby('Product Name')['Gross Profit'].sum()
ProductWise_Profit=ProductWise_Profit.sort_values(ascending=False).reset_index()
ProductWise_Profit.index=ProductWise_Profit.index +1
ProductWise_Profit

,Product Name,Gross Profit
1,Wonka Bar -Scrumdiddlyumptious,19357.50
2,Wonka Bar - Triple Dazzle Caramel,18610.20
3,Wonka Bar - Milk Chocolate,17443.37
4,Wonka Bar - Nutty Crunch Surprise,16819.95
5,Wonka Bar - Fudge Mallows,16593.60
6,Lickable Wallpaper,3930.00
7,Wonka Gum,310.70
8,Everlasting Gobstopper,104.00
9,Kazookles,92.75
10,Hair Toffee,59.50


In [41]:
## Rank products by:
# Gross Margin
ProductWise_Profit = df.groupby('Product Name')[['Sales','Gross Profit']].sum()
ProductWise_Profit['Gross Margin']=((ProductWise_Profit['Gross Profit']/ProductWise_Profit['Sales'])*100).round(2)
ProductWise_Profit=ProductWise_Profit.sort_values(by='Gross Margin',ascending=False).reset_index()
ProductWise_Profit.index=ProductWise_Profit.index +1
ProductWise_Profit

,Product Name,Sales,Gross Profit,Gross Margin
1,Everlasting Gobstopper,130.00,104.00,80.00
2,Hair Toffee,76.50,59.50,77.78
3,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,71.35
4,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,69.44
5,Wonka Bar - Fudge Mallows,24890.40,16593.60,66.67
6,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,65.33
7,Wonka Bar - Milk Chocolate,26867.75,17443.37,64.92
8,Laffy Taffy,53.73,33.48,62.31
9,Fizzy Lifting Drinks,78.75,47.25,60.00
10,Wonka Gum,597.50,310.70,52.00


• Identify:
○ High-profit / high-margin products
○ High-sales / low-margin products
○ Low-sales / low-profit products

In [43]:
def margin_status (x):
    if x['Profit_%'] > 50 & x['Profit_%'] > 50:
        return 'High Profit'
    elif x['Profit_%'] > 60:
        return 'Medium Profit'
    elif x['Profit_%'] > 40:
        return 'Low Profit'
    else:
        return 'Very Low Profit'

In [47]:
## Identify:
# High-profit / high-margin products
# High-sales / low-margin products
# Low-sales / low-profit products
    
product_rnk=df.groupby('Product Name')[['Gross Profit','Sales','Cost']].sum()
product_rnk['Profit_%']=((product_rnk['Gross Profit']/product_rnk['Sales'])*100).round(2)
product_rnk=product_rnk.sort_values(by='Gross Profit',ascending=False).reset_index()
product_rnk['Profit_Status']=product_rnk.apply(margin_status, axis=1)
product_rnk.index=product_rnk.index + 1
product_rnk

,Product Name,Gross Profit,Sales,Cost,Profit_%,Profit_Status
1,Wonka Bar -Scrumdiddlyumptious,19357.50,27874.80,8517.30,69.44,Medium Profit
2,Wonka Bar - Triple Dazzle Caramel,18610.20,28485.00,9874.80,65.33,Medium Profit
3,Wonka Bar - Milk Chocolate,17443.37,26867.75,9424.38,64.92,Medium Profit
4,Wonka Bar - Nutty Crunch Surprise,16819.95,23574.95,6755.00,71.35,Medium Profit
5,Wonka Bar - Fudge Mallows,16593.60,24890.40,8296.80,66.67,Medium Profit
6,Lickable Wallpaper,3930.00,7860.00,3930.00,50.00,Low Profit
7,Wonka Gum,310.70,597.50,286.80,52.00,Low Profit
8,Everlasting Gobstopper,104.00,130.00,26.00,80.00,Medium Profit
9,Kazookles,92.75,1205.75,1113.00,7.69,Very Low Profit
10,Hair Toffee,59.50,76.50,17.00,77.78,Medium Profit


--------------------------------------------------------------------------------------------------------------------------------------------------------

### Observations : Which products represent margin risk
1. Kazookles is the only product with a margin risk, having a profit margin of 7.69% while selling over 370 units.
2. Sweetarts and Nerds are the products with a profit margin of 46.67%.

### Profitability Metric Calculation For each product:
1. Gross Margin (%)
2. Profit per unit
3. Total profit contribution

In [23]:
top_profit=df.sort_values(by='Profit Contribution %',ascending=False)
plt.figure()
plt.pie('Profit Contribution %'.index, top_profit['Profit Contribution %'])
plt.xlabel('Profit Contribution %')
plt.ylabel('Product Name')
plt.title('Profit Contribution % Per Product')
plt.show()

KeyError: 'Profit Contribution %'

## Product-Level Profitability Analysis
Rank products by:
1. Gross profit
2. Gross margin

In [28]:
## Gross Profit Wise Product Rank
df=df.groupby('Product Name')[['Sales','Gross Profit']].sum()
Total_Profit=df['Gross Profit'].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
df=df.sort_values(by='Gross Profit', ascending=False).reset_index()
df.index=df.index +1
df

,Product Name,Sales,Gross Profit,Gross Margin
1,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,69.44
2,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,65.33
3,Wonka Bar - Milk Chocolate,26867.75,17443.37,64.92
4,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,71.35
5,Wonka Bar - Fudge Mallows,24890.40,16593.60,66.67
6,Lickable Wallpaper,7860.00,3930.00,50.00
7,Wonka Gum,597.50,310.70,52.00
8,Everlasting Gobstopper,130.00,104.00,80.00
9,Kazookles,1205.75,92.75,7.69
10,Hair Toffee,76.50,59.50,77.78


In [29]:
## Gross Margin Wise Product Rank
df=df.groupby('Product Name')[['Sales','Gross Profit']].sum()
Total_Profit=df['Gross Profit'].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
df=df.sort_values(by='Gross Margin', ascending=False).reset_index()
df.index=df.index +1
df

,Product Name,Sales,Gross Profit,Gross Margin
1,Everlasting Gobstopper,130.00,104.00,80.00
2,Hair Toffee,76.50,59.50,77.78
3,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,71.35
4,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,69.44
5,Wonka Bar - Fudge Mallows,24890.40,16593.60,66.67
6,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,65.33
7,Wonka Bar - Milk Chocolate,26867.75,17443.37,64.92
8,Laffy Taffy,53.73,33.48,62.31
9,Fizzy Lifting Drinks,78.75,47.25,60.00
10,Wonka Gum,597.50,310.70,52.00


## Identify:
1. High-profit / high-margin products
2. High-sales / low-margin products
3. Low-sales / low-profit products

In [38]:
## Def function for getting margin and sales status
def product_status(x):
    if (x['Gross Profit']) > 10000 and (x['Gross Margin'] > 50):
        return "High Profit & High Margin"
    elif (x['Sales'] > 5000) and (x['Gross Margin'] < 60) :
        return "High Sales & Low Margin"
    elif (x['Sales'] < 5000) and (x['Gross Profit'] > 1000) :
        return "Low Sales & Low Profit"
    else :
        return ("Moderate Performance")

In [43]:
Performance_Status=df.groupby('Product Name')[['Sales','Gross Profit']].sum()
Performance_Status['Gross Margin']=((Performance_Status['Gross Profit']/Performance_Status['Sales'])*100).round(2)
Performance_Status['product_status']=Performance_Status.apply(product_status, axis=1)
Performance_Status=Performance_Status.sort_values(by='Gross Profit', ascending=False).reset_index()
Performance_Status.index=Performance_Status.index +1
Performance_Status

,Product Name,Sales,Gross Profit,Gross Margin,product_status
1,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,69.44,High Profit & High Margin
2,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,65.33,High Profit & High Margin
3,Wonka Bar - Milk Chocolate,26867.75,17443.37,64.92,High Profit & High Margin
4,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,71.35,High Profit & High Margin
5,Wonka Bar - Fudge Mallows,24890.40,16593.60,66.67,High Profit & High Margin
6,Lickable Wallpaper,7860.00,3930.00,50.00,High Sales & Low Margin
7,Wonka Gum,597.50,310.70,52.00,Moderate Performance
8,Everlasting Gobstopper,130.00,104.00,80.00,Moderate Performance
9,Kazookles,1205.75,92.75,7.69,Moderate Performance
10,Hair Toffee,76.50,59.50,77.78,Moderate Performance


## Division-Level Performance Analysis
# Aggregate metrics by Division
# Compare:
1. Average margin by division
2. Revenue vs profit imbalance

In [51]:
## 1. Average margin by division
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
div_margin=df.groupby('Division')['Gross Margin'].mean().round(2)
div_margin

Division
Chocolate    67.46
Other        37.67
Sugar        57.69
Name: Gross Margin, dtype: float64

## 2. Revenue vs profit imbalance

In [58]:
def revenue_profit_imbalance(x):
    if (x['Sales'] > 20000) and (x['Gross Margin'] < 50):
        return "High Sales With Low Margin"
    elif (x['Sales'] < 10000) and (x['Gross Margin'] > 50):
        return "Low Sales With High Margin"
    elif (x['Sales'] < 2000) and (x['Gross Margin'] < 20):
        return "Low Sales With Low Margin"
    elif (x['Sales'] > 10000) and (x['Gross Margin'] > 50):
        return "High Sales With High Margin"
    else:
        return "Balanced"

In [61]:
margin_imbalance=df.groupby('Product Name')[['Sales','Gross Profit']].sum()
margin_imbalance['Gross Margin']=((margin_imbalance['Gross Profit']/margin_imbalance['Sales'])*100).round(2)
margin_imbalance['Imbalance Status']=margin_imbalance.apply(revenue_profit_imbalance, axis=1)
margin_imbalance=margin_imbalance.sort_values(by='Gross Profit', ascending=False).reset_index()
margin_imbalance.index=margin_imbalance.index +1
margin_imbalance

,Product Name,Sales,Gross Profit,Gross Margin,Imbalance Status
1,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,69.44,High Sales With High Margin
2,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,65.33,High Sales With High Margin
3,Wonka Bar - Milk Chocolate,26867.75,17443.37,64.92,High Sales With High Margin
4,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,71.35,High Sales With High Margin
5,Wonka Bar - Fudge Mallows,24890.40,16593.60,66.67,High Sales With High Margin
6,Lickable Wallpaper,7860.00,3930.00,50.00,Balanced
7,Wonka Gum,597.50,310.70,52.00,Low Sales With High Margin
8,Everlasting Gobstopper,130.00,104.00,80.00,Low Sales With High Margin
9,Kazookles,1205.75,92.75,7.69,Low Sales With Low Margin
10,Hair Toffee,76.50,59.50,77.78,Low Sales With High Margin


## Division-Level Performance Analysis
## Aggregate metrics by Division
# Compare:
# Identify divisions with:
1. Strong financial efficiency
2. Structural margin issues

In [136]:
## Def function for Financial Efficiency
def financial_efficiency(x):
    if (x['Gross Margin']>50):
        return "Strong Financial Efficiency"
    elif (x['Gross Margin']<20):
        return "Weak Financial Efficiency"
    else:
        return "Moderate"

In [139]:
total_profit=df['Gross Profit'].sum()
Fin_eff=df.groupby('Division')[['Sales','Gross Profit']].sum()
Fin_eff['Gross Margin']=((Fin_eff['Gross Profit']/Fin_eff['Sales'])*100).round(2)
Fin_eff['Efficiency Status']=Fin_eff.apply(financial_efficiency, axis=1)
Fin_eff['Profit_%']=(Fin_eff['Gross Profit']/total_profit*100).round(2)
Fin_eff=Fin_eff.sort_values(by='Sales', ascending=False).reset_index()
Fin_eff.index=Fin_eff.index +1
Fin_eff

,Division,Sales,Gross Profit,Gross Margin,Efficiency Status,Profit_%
1,Chocolate,131692.90,88824.62,67.45,Strong Financial Efficiency,95.06
2,Other,9663.25,4333.45,44.84,Moderate,4.64
3,Sugar,427.48,284.73,66.61,Strong Financial Efficiency,0.30


## ML PROJECT
Profit Concentration (Pareto) Analysis
• Determine % of products contributing:
○ 80% of revenue
○ 80% of profit
• Detect congestion-prone states or regionIdentify over-dependency risks

In [35]:
total_sales=df['Sales'].sum()
total_profit=df['Gross Profit'].sum()
product_category_profit=df.groupby('Product Name')[['Sales','Gross Profit']].sum().reset_index()
product_category_profit=product_category_profit.sort_values(by='Gross Profit', ascending=False)
product_category_profit['Cumulative Sales']=(product_category_profit['Sales']/total_sales*100).cumsum()
product_category_profit['Cumulative Profit']=(product_category_profit['Gross Profit']/total_profit*100).cumsum()
top_products=product_category_profit[product_category_profit['Cumulative Profit']<=80]
top_products

,Product Name,Sales,Gross Profit,Cumulative Sales,Cumulative Profit
13,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,19.660098,20.715882
12,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,39.750569,40.632023
10,Wonka Bar - Milk Chocolate,26867.75,17443.37,58.700394,59.299454
11,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,75.327808,77.299717


In [12]:
total_sales=df['Sales'].sum()
total_profit=df['Gross Profit'].sum()
product_category_profit=df.groupby('Region')[['Gross Profit','Sales']].sum()
product_category_profit=product_category_profit.sort_values(by='Sales', ascending=False)
product_category_profit['Sales_%']=(product_category_profit['Sales']/total_sales*100).round(2)
product_category_profit['Cumulative Sales']=(product_category_profit['Sales']/total_sales*100).cumsum()
product_category_profit['Cumulative Profit']=(product_category_profit['Gross Profit']/total_profit*100).cumsum()
product_category_profit['Total Sales']=total_sales
top_products=product_category_profit[product_category_profit['Cumulative Profit']<=101]
top_products

,Gross Profit,Sales,Sales_%,Cumulative Sales,Cumulative Profit,Total Sales
Region,,,,,,
Pacific,30485.94,46301.53,32.66,32.656471,32.625242,141783.63
Atlantic,26973.70,41197.24,29.06,61.712886,61.491779,141783.63
Interior,21282.49,32037.60,22.60,84.309007,84.267734,141783.63
Gulf,14700.67,22247.26,15.69,100.000000,100.000000,141783.63


In [6]:
region_sale=df.groupby('Region')['Sales'].sum()
region_sale

Region
Atlantic    41197.24
Gulf        22247.26
Interior    32037.60
Pacific     46301.53
Name: Sales, dtype: float64

Cost Structure Diagnostics
• Cost vs sales scatter analysis
• Identify:
○ Cost-heavy, margin-poor products
○ Pricing inefficiencies
• Flag products needing:
○ Repricing
○ Cost renegotiation
○ Discontinuation review

In [79]:
## Cost-heavy, margin-poor products
df=df.groupby('Product Name')[['Gross Profit','Sales']].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
cost_heavy_products=df[(df['Gross Margin']<50) & (df['Sales']>1000)]
cost_heavy_products

,Gross Profit,Sales,Gross Margin
Product Name,,,
Kazookles,92.75,1205.75,7.69


In [77]:
df=df.groupby('Product Name')[['Gross Profit','Sales']].sum()
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
df.sort_values(by='Gross Margin', ascending=False)

,Gross Profit,Sales,Gross Margin
Product Name,,,
Everlasting Gobstopper,104.00,130.00,80.00
Hair Toffee,59.50,76.50,77.78
Wonka Bar - Nutty Crunch Surprise,16819.95,23574.95,71.35
Wonka Bar -Scrumdiddlyumptious,19357.50,27874.80,69.44
Wonka Bar - Fudge Mallows,16593.60,24890.40,66.67
Wonka Bar - Triple Dazzle Caramel,18610.20,28485.00,65.33
Wonka Bar - Milk Chocolate,17443.37,26867.75,64.92
Laffy Taffy,33.48,53.73,62.31
Fizzy Lifting Drinks,47.25,78.75,60.00


In [83]:
## Pricing inefficiencies
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
pricing_issues=df[(df['Sales']>1000) & (df['Gross Margin']<20)]
pricing_issues

,Gross Profit,Sales,Gross Margin
Product Name,,,
Kazookles,92.75,1205.75,7.69


In [89]:
## Find sales greater than median
pricing_issues=df[(df['Sales']>df['Sales'].median()) & (df['Gross Margin'] < 60)]
pricing_issues.sort_values(by='Sales', ascending=False)

,Gross Profit,Sales,Gross Margin
Product Name,,,
Lickable Wallpaper,3930.00,7860.00,50.00
Kazookles,92.75,1205.75,7.69


In [47]:
total_sales = df['Sales'].sum()
total_profit = df['Gross Profit'].sum()

product_category = df.groupby('Product Name')[['Sales','Gross Profit']].sum().reset_index()

product_category = product_category.sort_values(by='Gross Profit', ascending=False)

product_category['Cumulative Sales'] = product_category['Sales'].cumsum()
product_category['Cumulative Sales %'] = (product_category['Cumulative Sales'] / total_sales * 100).round(2)

product_category['Cumulative Profit'] = product_category['Gross Profit'].cumsum()
product_category['Cumulative Profit %'] = (product_category['Cumulative Profit'] / total_profit * 100).round(2)

top_products = product_category[product_category['Cumulative Profit %'] <= 80]

top_products

,Product Name,Sales,Gross Profit,Cumulative Sales,Cumulative Sales %,Cumulative Profit,Cumulative Profit %
13,Wonka Bar -Scrumdiddlyumptious,27874.80,19357.50,27874.80,19.66,19357.50,20.72
12,Wonka Bar - Triple Dazzle Caramel,28485.00,18610.20,56359.80,39.75,37967.70,40.63
10,Wonka Bar - Milk Chocolate,26867.75,17443.37,83227.55,58.70,55411.07,59.30
11,Wonka Bar - Nutty Crunch Surprise,23574.95,16819.95,106802.50,75.33,72231.02,77.30


In [90]:
## Find sales greater than mean for repricing
pricing_issues=df[(df['Sales']>df['Sales'].mean()) & (df['Gross Margin'] < 60)]
pricing_issues.sort_values(by='Sales', ascending=False)

,Gross Profit,Sales,Gross Margin
Product Name,,,


In [48]:
## Cost renegotiation
df=df.groupby('Product Name')[['Cost','Sales']].sum()
df['Cost Ratio']=(df['Cost']/df['Sales']*100).round(1)
Cost_Renegotiation=df[df['Cost Ratio']>60]
Cost_Renegotiation

,Cost,Sales,Cost Ratio
Product Name,,,
Kazookles,1113.0,1205.75,92.3


In [123]:
## Discontinuation review
discontinue_product=df.groupby('Product Name')[['Sales','Gross Profit']].sum()
discontinue_product=discontinue_product[(discontinue_product['Sales']<1000) & (discontinue_product['Gross Profit'] <1000)]
discontinue_product

,Sales,Gross Profit
Product Name,,
Everlasting Gobstopper,130.00,104.00
Fizzy Lifting Drinks,78.75,47.25
Fun Dip,12.00,4.80
Hair Toffee,76.50,59.50
Laffy Taffy,53.73,33.48
Nerds,15.00,7.00
SweeTARTS,61.50,28.70
Wonka Gum,597.50,310.70


In [133]:
## Discontinuation review
df['Gross Margin']=((df['Gross Profit']/df['Sales'])*100).round(2)
discontinued_products=df[((df['Sales']>df['Sales'].median()) & (df['Gross Margin'] <10))]
discontinued_products.groupby('Product Name')[['Sales','Gross Profit']].sum()

,Sales,Gross Profit
Product Name,,
Kazookles,832.0,64.0


In [11]:
## City Wise Sales
city_wise=df.groupby('State/Province')['Sales'].sum()
city_wise.sort_values(ascending=False)

State/Province
California                   27917.40
New York                     15541.03
Texas                        13416.09
Pennsylvania                  8027.03
Washington                    6921.15
Illinois                      6898.96
Ohio                          6768.95
Florida                       4804.02
Arizona                       3587.55
North Carolina                3450.86
Michigan                      3331.00
Virginia                      3177.84
Georgia                       2692.84
Colorado                      2544.91
Tennessee                     2383.56
Indiana                       2002.78
Wisconsin                     1920.50
Oregon                        1920.15
Kentucky                      1880.63
Massachusetts                 1806.99
New Jersey                    1578.35
Maryland                      1476.99
Delaware                      1323.59
Minnesota                     1161.46
Alabama                        995.11
Connecticut                    980.

In [6]:
## Correlation between Gross Profit and Sales
profit_sales_corr=df[['Gross Profit','Sales']].corr()
profit_sales_corr

,Gross Profit,Sales
Gross Profit,1.000000,0.976404
Sales,0.976404,1.000000


In [10]:
df.corr(numeric_only=True)

,Row ID,Customer ID,Sales,Units,Gross Profit,Cost
Row ID,1.000000,-0.008749,0.005513,0.000502,0.008031,0.001812
Customer ID,-0.008749,1.000000,-0.005401,-0.011616,-0.005964,-0.004274
Sales,0.005513,-0.005401,1.000000,0.729347,0.976404,0.958986
Units,0.000502,-0.011616,0.729347,1.000000,0.815820,0.563344
Gross Profit,0.008031,-0.005964,0.976404,0.815820,1.000000,0.875144
Cost,0.001812,-0.004274,0.958986,0.563344,0.875144,1.000000


In [43]:
## Merging of 2 Tables
df_merged=pd.merge(df,df1, on='Division', how='left')
df_merged

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Country/Region,City,State/Province,Postal Code,Division,Region,Product ID,Product Name_x,Sales,Units,Gross Profit,Cost,Product Name_y,Factory
0,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.5,2,4.22,2.28,Wonka Bar - Nutty Crunch Surprise,Lot's O' Nuts
1,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.5,2,4.22,2.28,Wonka Bar - Fudge Mallows,Lot's O' Nuts
2,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.5,2,4.22,2.28,Wonka Bar -Scrumdiddlyumptious,Lot's O' Nuts
3,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.5,2,4.22,2.28,Wonka Bar - Milk Chocolate,Wicked Choccy's
4,1,US-2021-103800-CHO-MIL-31000,03-01-2024,30-06-2026,Standard Class,103800,United States,Houston,Texas,77095,Chocolate,Interior,CHO-MIL-31000,Wonka Bar - Milk Chocolate,6.5,2,4.22,2.28,Wonka Bar - Triple Dazzle Caramel,Wicked Choccy's
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
50695,10194,CA-2024-143500-CHO-SCR-58000,30-12-2025,26-06-2030,Standard Class,143500,Canada,Charlottetown,Prince Edward Island,C0A,Chocolate,Atlantic,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.8,3,7.50,3.30,Wonka Bar - Nutty Crunch Surprise,Lot's O' Nuts
50696,10194,CA-2024-143500-CHO-SCR-58000,30-12-2025,26-06-2030,Standard Class,143500,Canada,Charlottetown,Prince Edward Island,C0A,Chocolate,Atlantic,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.8,3,7.50,3.30,Wonka Bar - Fudge Mallows,Lot's O' Nuts
50697,10194,CA-2024-143500-CHO-SCR-58000,30-12-2025,26-06-2030,Standard Class,143500,Canada,Charlottetown,Prince Edward Island,C0A,Chocolate,Atlantic,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.8,3,7.50,3.30,Wonka Bar -Scrumdiddlyumptious,Lot's O' Nuts
50698,10194,CA-2024-143500-CHO-SCR-58000,30-12-2025,26-06-2030,Standard Class,143500,Canada,Charlottetown,Prince Edward Island,C0A,Chocolate,Atlantic,CHO-SCR-58000,Wonka Bar -Scrumdiddlyumptious,10.8,3,7.50,3.30,Wonka Bar - Milk Chocolate,Wicked Choccy's


In [26]:
## Factory Performance
factory_performance=df_merged.groupby('Factory')[['Sales','Gross Profit','Cost']].sum()
total_profit=df_merged['Gross Profit'].sum()
total_sales=df_merged['Sales'].sum()
factory_performance['Sales_%']=(factory_performance['Sales']/total_sales*100).round(2)
factory_performance['Profit_%']=(factory_performance['Gross Profit']/total_sales*100).round(2)
factory_performance['Gross Margin_%']=(factory_performance['Cost']/factory_performance['Sales']*100).round(2)
factory_performance

,Sales,Gross Profit,Cost,Sales_%,Profit_%,Gross Margin_%
Factory,,,,,,
Lot's O' Nuts,395078.70,266473.86,128604.84,56.47,38.08,32.55
Secret Factory,19753.98,8951.63,10802.35,2.82,1.28,54.68
Sugar Shack,11373.17,5472.37,5900.80,1.63,0.78,51.88
The Other Factory,10090.73,4618.18,5472.55,1.44,0.66,54.23
Wicked Choccy's,263385.80,177649.24,85736.56,37.64,25.39,32.55
